src/sql-parser/src/parser.rs

```diff
@@ -4837,7 +4837,80 @@ impl<'a> Parser<'a> {
     /// Parse an `EXPLAIN` statement, assuming that the `EXPLAIN` token
     /// has already been consumed.
     fn parse_explain_new(&mut self) -> Result<Statement<Raw>, ParserError> {
-        Err(ParserError::new(self.index, "unimplemented"))
+        let stage = match self.parse_one_of_keywords(&[
+            RAW,
+            DECORRELATED,
+            OPTIMIZED,
+            PHYSICAL,
+            OPTIMIZER,
+            QUERY,
+        ]) {
+            Some(RAW) => NewExplainStage::RawPlan,
+            Some(QUERY) => {
+                self.expect_keyword(GRAPH)?;
+                NewExplainStage::QueryGraph
+            }
+            Some(DECORRELATED) => NewExplainStage::DecorrelatedPlan,
+            Some(OPTIMIZED) => {
+                if self.parse_keyword(QUERY) {
+                    self.expect_keyword(GRAPH)?;
+                    NewExplainStage::OptimizedQueryGraph
+                } else {
+                    NewExplainStage::OptimizedPlan
+                }
+            }
+            Some(PHYSICAL) => NewExplainStage::PhysicalPlan,
+            Some(OPTIMIZER) => NewExplainStage::Trace,
+            None => NewExplainStage::OptimizedPlan,
+            _ => unreachable!(),
+        };
+        // TODO (#13137): Make specifying the format optional upon getting rid
+        // of the old explain syntax.
+        let format = match self.parse_one_of_keywords(&[TEXT, JSON]) {
+            Some(TEXT) => ExplainFormat::Text,
+            Some(JSON) => ExplainFormat::Json,
+            None => return Err(ParserError::new(self.index, "expected a format")),
+            _ => unreachable!(),
+        };
+
+        if stage == NewExplainStage::Trace {
```

Comment:
Ouch, I didn't realize that putting the format identifier before the noun would result in this code. The `parse_explain_old` nicely has `self.expect_keywords(&[...])?` in for each stage local to the if branch that is parsing this stage.